### PR TITLE
Reduce the snowflake async buffer size

### PR DESCRIPTION
**Pull request by benmoriceau**
## What
Reduce the async buffer size ratio for the snowflake destination from 0.7 to 0.5

### CODE REVIEW

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/StagingConsumerFactory.java [hunk 1]
```diff
@@ -101,6 +102,34 @@ public SerializedAirbyteMessageConsumer createAsync(final Consumer<AirbyteMessag
...
+                                                      final NamingConventionTransformer namingResolver,
+                                                      final JsonNode config,
+                                                      final ConfiguredAirbyteCatalog catalog,
+                                                      final boolean purgeStagingData,
+                                                      final TypeAndDedupeOperationValve typerDeduperValve,
+                                                      final TyperDeduper typerDeduper,
+                                                      final ParsedCatalog parsedCatalog,
+                                                      final String defaultNamespace,
+                                                      final boolean useDestinationsV2Columns,
+                                                      Optional<Long> BufferMemoryLimit) {
```
Comment (tryangul) [inline airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/StagingConsumerFactory.java:132]: needs final. Also casing

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/StagingConsumerFactory.java [hunk 2]
```diff
@@ -112,7 +141,7 @@ public SerializedAirbyteMessageConsumer createAsync(final Consumer<AirbyteMessag
         () -> GeneralStagingFunctions.onCloseFunction(database, stagingOperations, writeConfigs, purgeStagingData, typerDeduper).accept(false),
         flusher,
         catalog,
-        new BufferManager(),
+        BufferMemoryLimit.map(memoryLimit -> new BufferManager(memoryLimit)).orElse(new BufferManager()),
```
Comment (tryangul) [inline airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/StagingConsumerFactory.java:144]: can we put this in a variable or method for clarity

Something like 
`bufferMemoryLimit.getOrElse(DEFAULT_VALUE)`

**Review by akashkulk — COMMENTED** [review]
Just trying to understand the tradeoffs here : does this mean that the destination will be slower assuming the source is already producing records faster than the destination can consume?

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`b4aeb3f022`](https://github.com/airbytehq/airbyte/pull/30671/commits/b4aeb3f022b5fd7133250c4e461d524441e8e380)) - ❌

⏲️  Total pipeline duration: 77mn34s 

| Step                                                                         | Result   |
|:-----------------------------------------------------------------------------|:---------|
| Build connector tar                                                          | ✅        |
| Build destination-snowflake docker image for platform linux/x86_64           | ✅        |
| Java Connector Unit Tests                                                    | ✅        |
| Java Connector Integration Tests                                             | ❌        |
| Validate airbyte-integrations/connectors/destination-snowflake/metadata.yaml | ✅        |
| Connector version semver check                                               | ✅        |
| Connector version increment check                                            | ✅        |
| QA checks                                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/bmoric_reduce-snowflake-buffer-size/1695319223/b4aeb3f022b5fd7133250c4e461d524441e8e380/destination-snowflake/3.1.12/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:b4aeb3f022b5fd7133250c4e461d524441e8e380)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/StagingConsumerFactory.java (hunk 1 above)
Comment (benmoriceau) [reply airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/StagingConsumerFactory.java]: done

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/StagingConsumerFactory.java (hunk 2 above)
Comment (benmoriceau) [reply airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/StagingConsumerFactory.java]: Done

**Comment by benmoriceau** [comment]
> Just trying to understand the tradeoffs here : does this mean that the destination will be slower assuming the source is already producing records faster than the destination can consume?

Not exactly, we will still buffer the main difference is that we will reach the limit of the buffer faster (because it is too small). In most of those case, we would have reach the buffer limit even if we had a bigger buffer, reducing the size will just makes use reaching the limit faster.

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`5ec7b96795`](https://github.com/airbytehq/airbyte/pull/30671/commits/5ec7b967954d56378dc6d3d5dfe1cc7256804394)) - ❌

⏲️  Total pipeline duration: 78mn16s 

| Step                                                                         | Result   |
|:-----------------------------------------------------------------------------|:---------|
| Build connector tar                                                          | ✅        |
| Build destination-snowflake docker image for platform linux/x86_64           | ✅        |
| Java Connector Unit Tests                                                    | ✅        |
| Java Connector Integration Tests                                             | ❌        |
| Validate airbyte-integrations/connectors/destination-snowflake/metadata.yaml | ✅        |
| Connector version semver check                                               | ✅        |
| Connector version increment check                                            | ✅        |
| QA checks                                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/bmoric_reduce-snowflake-buffer-size/1695323980/5ec7b967954d56378dc6d3d5dfe1cc7256804394/destination-snowflake/3.1.12/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:5ec7b967954d56378dc6d3d5dfe1cc7256804394)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by akashkulk** [comment]
> > Just trying to understand the tradeoffs here : does this mean that the destination will be slower assuming the source is already producing records faster than the destination can consume?
> 
> Not exactly, we will still buffer the main difference is that we will reach the limit of the buffer faster (because it is too small). In most of those case, we would have reach the buffer limit even if we had a bigger buffer, reducing the size will just makes use reaching the limit faster.

And just so that I understand - this will just result in more frequent uploads to the destination (Snowflake)?

Approving from my end - but will defer to @tryangul for a more thorough review

**Review by akashkulk — APPROVED** [review]

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`f22c6c0fa9`](https://github.com/airbytehq/airbyte/pull/30671/commits/f22c6c0fa975b5d6d4618384ca13773169124811)) - ✅

⏲️  Total pipeline duration: 13mn50s 

| Step                                                                         | Result   |
|:-----------------------------------------------------------------------------|:---------|
| Build connector tar                                                          | ✅        |
| Build destination-snowflake docker image for platform linux/x86_64           | ✅        |
| Java Connector Unit Tests                                                    | ✅        |
| Java Connector Integration Tests                                             | ✅        |
| Validate airbyte-integrations/connectors/destination-snowflake/metadata.yaml | ✅        |
| Connector version semver check                                               | ✅        |
| Connector version increment check                                            | ✅        |
| QA checks                                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/bmoric_reduce-snowflake-buffer-size/1695328847/f22c6c0fa975b5d6d4618384ca13773169124811/destination-snowflake/3.1.12/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:f22c6c0fa975b5d6d4618384ca13773169124811)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by tryangul** [comment]
> And just so that I understand - this will just result in more frequent uploads to the destination (Snowflake)?

It just means we will buffer less before backpressuring. The goal is to buffer enough so our upload threads always have work available as soon as they finish, but not so much we oom. The original limits as they've been set were just best guess when this was originally designed.

**Review by tryangul — APPROVED** [review]

**Comment by octavia-approvington** [comment]
Myomoto says it looks good <br/> thats a niiice

**Review by octavia-approvington — APPROVED** [review]

**Pull request merged**